Lighting/Src/lighting_controller.cpp

```diff
@@ -212,40 +159,289 @@ void LightingController::remove_led_from_cd(uint8_t index,
+	this->domain_brightness[domain] = brightness;
+}
+
+void LightingController::set_domain_colour(ControlDomain domain, RGB_colour_t colour) {
+	this->domain_colours[domain] = colour;
+}
+
+void LightingController::activate_domain(ControlDomain domain) {
+	if (this->domain_allowed & (1 << domain)) {
+		this->domain_active |= 1 << domain;	// TODO: have active/disabled booleans
+		for (int i = 0; i < CD_LENGTH; ++i) {
+			if (this->domain_active & (1 << i)) { // IF THIS DOMAIN IS ACTIVE
+				for (int j = 0; j < NUM_LEDS; ++j) {
+					if (this->domain_leds[i] & (1 << j)) {
+						this->leds[j].set_led_colour(domain_colours[i],
+								domain_brightness[i]);
+					}
 				}
 			}
 		}
+	} else {
+		// Domains should not be active when they are disallowed.
+		// Instead of checking this every time we poll domains, check it on domain activation request.
+		this->domain_active &= ~(1 << domain);
 	}
 }
 
-void LightingController::disable_control_domain(ControlDomain domain) {
-	this->domain_state[domain] = false;
+
+void LightingController::deactivate_domain(ControlDomain domain) {
+	this->domain_active &= ~(1 << domain);
 	for (int i = 0; i < CD_LENGTH; ++i) {
-		if (this->domain_state[domain]) { // IF THIS DOMAIN IS ACTIVE
+		if (this->domain_active & (1 << i)) { // IF THIS DOMAIN IS ACTIVE
 			for (int j = 0; j < NUM_LEDS; ++j) {
 				if (this->domain_leds[i] & (1 << j)) {
-					this->leds[i].set_led_colour(domain_colours[i],
+					this->leds[j].set_led_colour(domain_colours[i],
 							domain_brightness[i]);
 				}
 			}
 		}
 	}
 }
 
+void LightingController::allow_domain(ControlDomain domain) {
+	this->domain_allowed |= 1 << domain;
+}
+
+void LightingController::disallow_domain(ControlDomain domain) {
+	this->domain_allowed &= ~(1 << domain);
+}
+
+
+void LightingController::add_leds_to_cd(ControlDomain domain, uint16_t leds) {
+	for (int i = 0; i < NUM_LEDS; ++i) {
+		if (leds & (1 << i)) {				//if this LED is one I want to add to the control domain.
+			add_led_to_cd(i, domain);
+		}
+	}
+}
+
+void LightingController::remove_leds_from_cd(ControlDomain domain, uint16_t leds) {
+	for (int i = 0; i < NUM_LEDS; ++i) {
+		if (leds & (1 << i)) {				//if this LED is one I want to remove from the control domain.
+			remove_led_from_cd(i, domain);
+		}
+	}
+}
+
+////////////////////
+// STATE MACHINE FN
+////////////////////
+void LightingController::transition_to(uint8_t input) {
+	uint16_t beacon_leds = 0;
+	uint16_t taxi_leds = (1 << 7) + (1 << 9);
+	uint16_t strobe_leds = 0;
+	uint16_t navigation_leds = 0;
+	uint16_t landing_leds = (1 << 1) + (1 << 2) + (1 << 3) + (1 << 4);
+	uint8_t search_leds = 63; //leds 0 to 5
+
+	switch (this->drone_state) {
+		case GROUND_STATE:
+			if (input == 0) {												//go into search state
+				//disable ALL possibly active beacon and strobe lights.
+				beacon_leds = 63;
+				strobe_leds = 56;
+				remove_leds_from_cd(CD_BEACON, beacon_leds);
+				remove_leds_from_cd(CD_BEACON, strobe_leds);
```

Comment:
This applies to all of the cases below, we don't need to remove leds from our command domain.

All we need to do is set the bit corresponding to the domain in `allowed_domains` to be 0 (this disables it from ever being on).

This does mean that we could possibly want a `domain_override` field that a user could use to override domains, but we can add this in a later PR

Nobody other than our code touches the active / non-active domains, and in the nominal case (non-debug), the only code changing active states is this state machine, so there's no possibility of a domani being accidentally enabled unless we got the code wrong, in which case adding/removing LED's isn't going to help us.